pkg/controller/trainjob_controller.go

```diff
@@ -105,11 +107,29 @@ func (r *TrainJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 func (r *TrainJobReconciler) reconcileObjects(ctx context.Context, runtime jobruntimes.Runtime, trainJob *trainer.TrainJob) (objsOpState, error) {
 	log := ctrl.LoggerFrom(ctx)
 
-	objs, err := runtime.NewObjects(ctx, trainJob)
+	objects, err := runtime.NewObjects(ctx, trainJob)
 	if err != nil {
 		return buildFailed, err
 	}
-	for _, obj := range objs {
+	for _, object := range objects {
+		// TODO (astefanutti): Remove conversion to unstructured when the runtime.ApplyConfiguration
+		//  interface becomes available and first-class SSA method is added to the controller-runtime
+		// client. See https://github.com/kubernetes/kubernetes/pull/129313
+		var obj client.Object
+		if o, ok := object.(client.Object); ok {
```

Comment:
If we need to keep object unstructured, do we really want to have SSA to reconcile TrainJob's objects ?
Can we just keep the existing way to create Kubernetes objects to reduce complexity ?

Reply:
Using unstructured here is more an implementation detail, and should not be what drives the decision.

Using SSA actually removes some complexity and brings some benefits w.r.t. to conflicts handing and fields ownership, which I think is why @tenzen-y created #2297.

Unstructured is needed here only because the controller-runtime client requires a client.Object at the moment but that's is going to be alleviated, and should not prevent us from adopting SSA now.

Reply:
Does it mean that all objects that users want to create with TrainJob should support ApplyConfiguration ?
For example, if tomorrow we will introduce a new SlurmRuntime which uses NodeSet instead of JobSet in the Job template, it should support SSA and ApplyConfiguration ?

Reply:
That makes sense.